### PR TITLE
#771: allow host header override

### LINKED ISSUE #771: BackendService override host header option
In some use cases, it would be desirable to be able to route a request with its Host header set to the value of the hostname property of the selected origin. This could be configured by an additional property on the BackendService (boolean overrideHostHeader). Current behavior maintains the host header of the incoming request, and this should still be the default behavior after this option is added.

### PR DESCRIPTION
**Pull request by ahsansyed97**
fixes #771

### CODE REVIEW

File: components/api/src/main/java/com/hotels/styx/api/extension/service/BackendService.kt
Comment (ahsansyed97) [inline components/api/src/main/java/com/hotels/styx/api/extension/service/BackendService.kt:22]: No importing *!

File: components/api/src/main/java/com/hotels/styx/api/extension/service/BackendService.kt
Comment (ryanellison) [inline components/api/src/main/java/com/hotels/styx/api/extension/service/BackendService.kt:39]: I believe we need to allow SSL settings to be null for non-secure services.

File: components/api/src/main/java/com/hotels/styx/api/extension/service/BackendService.kt
Comment (ryanellison) [inline components/api/src/main/java/com/hotels/styx/api/extension/service/BackendService.kt:152]: Yes, related to my other comment, this can be null.

File: components/api/src/main/java/com/hotels/styx/api/extension/service/BackendService.kt
Comment (ryanellison) [inline components/api/src/main/java/com/hotels/styx/api/extension/service/BackendService.kt:77]: Can this ever be null? If not, adjust type (similar comment for others in the builder, here)

File: components/api/src/main/java/com/hotels/styx/api/extension/service/BackendService.kt
Comment (ryanellison) [inline components/api/src/main/java/com/hotels/styx/api/extension/service/BackendService.kt:83]: If HealthCheckConfig isn't nullable on BackendService, this is sort of a problem if they don't set this, right? The solution might be to make this nullable on BackendService.

File: components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt
Comment (ryanellison) [inline components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt:169]: probably use
```
LiveHttpRequest =
...
```
instead of 
```
LiveHttpRequest {
...
```
where possible

File: components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt
Comment (ryanellison) [inline components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt:231]: un-Java-tize this.

File: components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt
Comment (ryanellison) [inline components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt:322]:
```
override fun avoidOrigins(): List<Origin> = emptyList()
```

File: components/api/src/main/java/com/hotels/styx/api/extension/service/BackendService.kt
Comment (yashvyas) [inline components/api/src/main/java/com/hotels/styx/api/extension/service/BackendService.kt:27]: consider making this a data class. seems to me like it just holds data.

File: components/api/src/main/java/com/hotels/styx/api/extension/service/BackendService.kt
Comment (yashvyas) [inline components/api/src/main/java/com/hotels/styx/api/extension/service/BackendService.kt:160]: i think this might be the same as the generated equals method on the data class. Pretty sure it compares each property, but maybe reviewing the generated implementation will tell us whether we need to override this (and others like hashCode) at all.

Comment (yashvyas) [reply components/api/src/main/java/com/hotels/styx/api/extension/service/BackendService.kt:168]: For toString it seems like the choice of curl braces vs parenthesis is the difference.

File: components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt
Comment (yashvyas) [inline components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt:90]: This should probably move to the top.

Comment (yashvyas) [reply components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt:90]: do review: https://kotlinlang.org/docs/coding-conventions.html#class-layout, seems like other changes are required. Just makes it easier to read code if everything is organized in the same way. For example, i was wondering why the class doesn't have the 
```
private val rewriteRuleset ...
```
until i reached this line.

File: components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt
Comment (yashvyas) [inline components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt:135]: too verbose, newPreviousOrigins = mutableListOf() works yeah?

File: components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt
Comment (yashvyas) [inline components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt:139]: too verbose, don't need to specify types here. Using named argument or the implicit "it" seems like a pretty subjective thing so i would follow project conventions there.

File: components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt
Comment (yashvyas) [reply components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt:135]: maybe this doesn't need to be mutable either.. do verify and consider.

File: components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt
Comment (yashvyas) [inline components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt:227]: some of these are still just single expressions that don't require a method body. do review.

File: components/proxy/src/main/java/com/hotels/styx/proxy/StyxBackendServiceClientFactory.kt
Comment (yashvyas) [inline components/proxy/src/main/java/com/hotels/styx/proxy/StyxBackendServiceClientFactory.kt:54]: seems odd with having just 1 argument, maybe it can be one line?

**Comment by yashvyas** [comment]
1. Use the public github issue in the PR title and commit message, this will link the change to the issue appropriately. Internal ticket is not going to be reference-able. 
2. Doc update should be included in this commit.

File: components/proxy/src/main/kotlin/com/hotels/styx/services/OriginsConfigConverter.kt
Comment (yashvyas) [inline components/proxy/src/main/kotlin/com/hotels/styx/services/OriginsConfigConverter.kt:101]: is there a good reason to change this signature? Why is it okay to take in a nullable healthCheckConfig here now?

Comment (yashvyas) [reply components/proxy/src/main/kotlin/com/hotels/styx/services/OriginsConfigConverter.kt:101]: better to use this method instead of the currently defined one? 
```
    internal fun healthCheckServices(apps: List<BackendService>) = apps
            .filter(::isHealthCheckConfigured)
            .mapNotNull {
                it.healthCheckConfig()?.let { hc ->
                    it.id() to hc
                }
            }
            .map { (id, healthCheckConfig) ->
                val appId = id.toString()
                val serviceName = "$appId-monitor"
                
                Pair(serviceName, healthCheckService(serviceName, appId, healthCheckConfig))
            }
```

**Review by kvosper — CHANGES_REQUESTED** [review]

File: components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt
Comment (ryanellison) [inline components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt:64]: Do you think we should add a check in .build() for loadBalancer not being null, as you did for metrics? Not a huge fan of using !! in production code, in any case...
You could potentially do like
```
originStatsFactory = builder?.originStatsFactory ?: somethingElseHere
```
even though you know that should never happen.

Comment (ahsansyed97) [reply components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt:64]: @ryanellison yeah I could definitely do that; @kvosper I'll tag you to get your thoughts too since you suggested the !!

I already check metrics and originStatsFactory being null in .build(), do you think I should also add a check for loadBalancer there **and** do something like your suggestion for **all three** of metrics, originStatsFactory, and loadBalancer here in the constructor to remove the need of !!?

Comment (ryanellison) [reply components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt:64]: I guess if @kvosper is alright with !! I'd just add an additional check into .build() to protect against this being null, since we're saying it's definitely not with !!.
The tricky think about providing a default for originStatsFactory is the the factory for one of those requires non-null "metrics" which starts to get a little ugly putting it all here...
For reference, in your builder:
```
OriginStatsFactory.CachingOriginStatsFactory(metrics)
```

**Comment by ryanellison** [comment]
Looks good to me, other than the comment added around the builder. @kvosper do we have the ability to release a snapshot of this for use in a CGP canary before merging?

File: components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt
Comment (ahsansyed97) [reply components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt:64]: Cool, I'll add a null check in .build() for loadBalancer and leave the !! as is for now then

Comment (ahsansyed97) [reply components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt:64]: ^ pushed changes, squashed into the previous commit

**Review by ryanellison — APPROVED** [review]
Approving, but let's wait for an approval from Kyle before merging.

File: components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt
Comment (kvosper) [reply components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt:64]: After thinking about it, `!!` is not the best choice, as this simply throws a `NullPointerException` if something is null. It would be better to use Kotlin's `requireNotNull` like so `requireNotNull(builder.loadBalancer) { "loadBalancer is required" } `.

I may have confused this with java's `Objects.requireNonNull` earlier (which should not be used in Kotlin), but using Kotlin's functions is OK.

**Review by kvosper — APPROVED** [review]

**Review by yashvyas — APPROVED** [review]

**Pull request merged**
